ouroboros-consensus/src/Ouroboros/Consensus/Node/Serialisation.hs

```diff
@@ -71,21 +72,21 @@ class SerialiseNodeToNode blk a where
 -- | Serialise a type @a@ so that it can be sent across the network via
 -- node-to-client protocol.
 class SerialiseNodeToClient blk a where
-  encodeNodeToClient :: CodecConfig blk -> BlockNodeToClientVersion blk -> a -> Encoding
-  decodeNodeToClient :: CodecConfig blk -> BlockNodeToClientVersion blk -> forall s. Decoder s a
+  encodeNodeToClient :: CodecConfig blk -> QueryVersion -> BlockNodeToClientVersion blk -> a -> Encoding
+  decodeNodeToClient :: CodecConfig blk -> QueryVersion -> BlockNodeToClientVersion blk -> forall s. Decoder s a
```

Comment:
Hmmm. A scenario occurred to me that is more concrete than what I had been thinking about so far.

Suppose we add the `GetPartialLedgerConfig` query in `QueryV_1`. Today, that query would return:

```
data ShelleyPartialLedgerConfig era = ShelleyPartialLedgerConfig {
      shelleyLedgerConfig    :: !(ShelleyLedgerConfig era)
    , shelleyTriggerHardFork :: !TriggerHardFork
    }

data ShelleyLedgerConfig era = ShelleyLedgerConfig {
      shelleyLedgerCompactGenesis     :: !(CompactGenesis era)
    , shelleyLedgerGlobals            :: !SL.Globals
    , shelleyLedgerTranslationContext :: !(Core.TranslationContext era)
    }
```

But note that we added the `shelleyLedgerTranslationContext` field as part of Alonzo. So that serialization would have had to change different if this top-level query had existed before we integrated Alonzo.

Does the `BlockNodeToClientVersion blk` argument to `encodeNodeToClient` provide enough information for us to have done that? Or would `QueryVersion` be relevant somehow?

More generally, what exactly does `QueryVersion` constrain/reflect? And how does its evolution relate to the evolution of `BlockNodeToClientVersion`?

I'm hoping you already have more clarity on this than I do and can share it with me, since it feels too nebulous right now.

Reply:
My understanding is that we DO aim to support clients that are on an older `NodeToClientVersion` than the node. Hence the existence of `NodeToClientVersion` does make sense.

In practice the client supports a single version and hopes that the node is backwards compatible with it. Often the node *is* backwards compatible. For example, going from version 3 to 4 we've added some new queries. In a way, this seems incompatible because the client can't decode responses to the newer queries. Actually, this *is* compatible because the older client will never use any of the new queries and so will never need to decode responses to them. The fact that these things are compatible requires understanding of how the protocol works. We don't get static checks for this compatibility i.e. via some sort of "migration" code. The lack of static checks is what makes things confusing. At least we have some golden tests to check that our serialization code matches serializations from older versions though I doubt this implies much confidence that our versioning code is correct.

Reply:
On that note, we already, in a way, allow backwards compatibility in the other direction: "newer" clients are compatible  with "older" nodes, but only if the newer client uses an older `NodeToClientVersion`. In that case it is that client's responsibility to avoid all functionality of newer versions even if the code (e.g. new query constructors) is available.

Reply:
Suppose a year from now we introduce a new era.

  - Release X: we have the `GetLedgerConfig :: Query (LedgerConfig blk) ` query  and the last era is `Foo`.
  - Release X+1: we still have the `GetLedgerConfig :: Query (LedgerConfig blk)` query  and the last era is `Bar`, the successor of `Foo`.

If the client is running X+1 and the server is running X, then the server will send the serialization of a `LedgerConfig blk` that excludes `Bar` but the client will try to deserialize that into a `LedgerConfig blk` that includes `Bar`.

I don't see how the newer client in that case could (easily & cleanly) avoid the difference other than not sending the `GetLedgerConfig` query, which seems too severe.

Reply:
I just looked over the types of the existing Shelley queries, wondering "how do they avoid the product semantics challenge?"

https://github.com/input-output-hk/ouroboros-network/blob/e50613562d6d4a0f933741fcf590b0f69a1eda67/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs#L82

All of the result types involve a _single_ `era`. So I think they all follow the sum semantics, not the product semantics. ... Maybe for `GetGenesisConfig` we're still going to have to figure out what to do for the Alonzo bit? I'm not quite sure what `GetGenesisConfig` is used for, and whether it needs to include the new Alonzo "genesis data".

Reply:
We had a call that included Duncan. In particular, he clarified that we do intend for newer nodes to support older clients, for the convenience of users doing incremental upgrades (updating the node first, since that's more important).

We three also concluded that as long as we avoid product semantics in the query arguments, this should be fine; it's a continuation of our practices so far: eg the node would truncate its HFC era list when serializing query results according to the negotiated `NodeToClientVersion`.

We do not anticipate _needing_ query with HFC-product arguments. But if we're wrong about that, we do have three main options in mind.

  - Option 1: decompose the hypothetical query into queries with a sum argument and therefore force the client to send multiple queries and assemble their results as it sees fit
  
  - Option 2: have the node deserializer fill-in the missing product components. By a parametricity-like argument, the fill value shouldn't matter. (We're not excited by this option.)
  
  - Option 3: disallow the query until the client catches up to the node, enforced via the negotiated version. (We're also not excited about this, since we generally don't want a node update to cause feature degradation in a client --- it's unlikely to be graceful degradation.)